File: Ts/src/AlternateText.ts
module SpinTexter {

    export class AlternatedText extends Array<ITextPart> implements ITextPart {

        private _rnd: Random;
        private _level: number;

        constructor(rnd: Random, level : number = 0) 
        {
            super();
            this._rnd = rnd != null ? rnd : new Random();
            this._level = level;
        }

        public toString(): string
        {
            if (this.length === 0) return ""; 
            var chosedIdx: number = this._rnd.nextInt(0, this.length -1);
            var result = this[chosedIdx].toString()
            return result;
        } 

        public toStructuredString(): string
        {
            var sb: StringBuilder = new SpinTexter.StringBuilder();
            var first: boolean = true;
            sb.append("{");
            this.forEach(function (el) {
                if (first) first = false;
                else sb.append("|"); 
                sb.append(el.toStructuredString());
            });
            sb.append("}");
            return sb.toString();
        } 
       
        // experimenal, not part of the interface - thinking how to expres shit as json???
        //================================================================================
        public toArrayString(): Array<any> 
        {
            //var sb: StringBuilder = new SpinTexter.StringBuilder();
            //var first: boolean = true;
            //if(this._level === 0)
            //    sb.append("[");
            //else
            //    sb.append("\", [");

            //this.forEach(function (el)
            //{
            //    if (first)
            //    {
            //        sb.append("\"");
            //    }
            //    else
            //        sb.append("\", \"");

            //    sb.append(el.toArrayString());

            //    if (first) {
            //        sb.append("\"");
            //    }

            //    if (first) 
            //        first = false;
                
            //});
            //sb.append("]");
            //return sb.toString();

            return this;

        }

        public countVariants(): number
        {
            //var res: number = 0;
            //this.forEach(function (tp) { res += tp.countVariants(); });
            //return res;
            var res: number = 1;
            this.forEach(function (tp) {
                var counter: number = tp.countVariants();
                if (res < 100100) res += counter;
            });
            return res;
        }

        // todo: make a guy that work for both of these guys...

        public countMinWords(): number {
            var counter = 0;
            this.forEach(function (tp) {                    
                var minWords = tp.countMinWords();
                if (counter === 0)
                    counter = minWords;
                if (minWords < counter)
                    counter = minWords;
            });
            return counter;
        }

        public countMaxWords(): number {
            var counter = 0;
            this.forEach(function (tp) {
                var maxWords = tp.countMinWords();
                if (counter === 0)
                    counter = maxWords;
                if (maxWords > counter)
                    counter = maxWords;
            });
            return counter;
            //return Math.max.apply(null, this);
        }
    }
}

File: Cs/AlternateText.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace HalloJoe.Spin
{
    /// <summary>
    /// Represents a textpart which is [random] one among its contained textparts
    /// </summary>
    internal class AlternatedText : List<ITextPart>, ITextPart
    {
        private Random _rnd;
        private ParserConfiguration _cfg;

        /// <summary>
        /// Constructor that take random or create new
        /// </summary>
        public AlternatedText(Random rnd = null, ParserConfiguration cfg = null) 
        {
            _cfg = cfg ?? new ParserConfiguration();
            _rnd = rnd ?? new Random();
        }
        /// <summary>
        /// Result string
        /// </summary>
        /// <returns></returns>
        string ITextPart.ToString()
        {
            if (this.Count == 0)
                return string.Empty;
            int chosedIdx = _rnd.Next(this.Count);
            return this[chosedIdx].ToString();
        }
        /// <summary>
        /// Structured string
        /// </summary>
        /// <returns></returns>
        string ITextPart.ToStructuredString()
        {
            StringBuilder sb = new StringBuilder();
            bool first = true;
            sb.Append(_cfg.Opening);
            foreach (ITextPart el in this)
            {
                if (first) first = false;
                else
                    sb.Append(_cfg.Delimiter);
                sb.Append(el.ToStructuredString());
            }
            sb.Append(_cfg.Closing);
            return sb.ToString();
        }

        //List<ITextPart> ITextPart.ToObject()
        //{
        //    return this;
        //}

        /// <summary>
        /// Count posible variants
        /// </summary>
        /// <returns></returns>
        decimal ITextPart.CountVariants()
        {
            decimal res = 0;
            foreach (ITextPart tp in this)
                res += tp.CountVariants();
            return res;
        }
        /// <summary>
        /// Count minimum words structured string can consist of
        /// </summary>
        /// <returns></returns>
        int ITextPart.CountMinWords() => this.Min(x => x.CountMinWords());
        /// <summary>
        /// Count maximum words structured string can consist of
        /// </summary>
        /// <returns></returns>
        int ITextPart.CountMaxWords() => this.Max(x => x.CountMaxWords());
    }
}
